Bluetooth: Controller: Fix hang due to loop in node_rx list

cvinayak opened the pull request:
Fix Controller hang due to infinite looping caused by
duplicate node_rx enqueued in auxiliary context.

When LLL scheduling is not applied due to invalid aux offset
then ULL scheduling too would skip setting up the reception
after similarly checking the validity of aux offset required
to schedule the reception of auxiliary PDU. This check in
ULL was after the received node_rx being enqueued into the
auxiliary context causing a loop in the list of node_rx.

Dequeue of a list with a loop of node_rx caused an infinite
loop execution in the Controller.

Regression introduced in commit https://github.com/zephyrproject-rtos/zephyr/commit/3590bd648f8fe39b04d81ee6431037e05f026406 ("Bluetooth:
Controller: Fix missing invalid aux offset check").

cvinayak committed:
Bluetooth: Controller: Infinite loop assertion on node_rx release

Add assertion check to catch infinite loop due to incorrect
node_rx release.

If same node_rx is released twice then it can lead to
infinite looping when releaseing link or node_rx buffers.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Fix hang due to loop in node_rx list

Fix Controller hang due to infinite looping caused by
duplicate node_rx enqueued in auxiliary context.

When LLL scheduling is not applied due to invalid aux offset
then ULL scheduling too would skip setting up the reception
after similarly checking the validity of aux offset required
to schedule the reception of auxiliary PDU. This check in
ULL was after the received node_rx being enqueued into the
auxiliary context causing a loop in the list of node_rx.

Dequeue of a list with a loop of node_rx caused an infinite
loop execution in the Controller.

Regression introduced in commit 3590bd648f8f ("Bluetooth:
Controller: Fix missing invalid aux offset check").

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak commented:
@mtpr-ot @Tronil @wopu-ot FYI.

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_scan_aux.c:
```diff
@@ -511,6 +512,22 @@ void ull_scan_aux_setup(memq_link_t *link, struct node_rx_pdu *rx)
+		window_size_us = OFFS_UNIT_30_US;
+	}
+
+	/* Calculate received aux offset we need to have ULL schedule a reception */
+	aux_offset_us = (uint32_t)PDU_ADV_AUX_PTR_OFFSET_GET(aux_ptr) * window_size_us;
```
Here you multiple 2 `uint32_t` and store the result in another `uint32_t`. Should have some sort of overflow protection

Thalley replied:
The typecast sounds OK but why is `window_size_us` defined as a `uint32_t` then? 

Since `window_size_us` is either `OFFS_UNIT_300_US` or `OFFS_UNIT_30_US`, then it looks like it can be a `uint16_t` and thus the line can become
```suggestion
	/* Calculate received aux offset we need to have ULL schedule a reception */
	aux_offset_us = ((uint16_t)PDU_ADV_AUX_PTR_OFFSET_GET(aux_ptr)) * window_size_us;
```
or maybe even just
```suggestion
	/* Calculate received aux offset we need to have ULL schedule a reception */
	aux_offset_us = PDU_ADV_AUX_PTR_OFFSET_GET(aux_ptr) * window_size_us;
```

It's a minor detail, and if you don't think `window_size_us` should be a `uint16_t`, then let me know and I can approve

cvinayak replied:
Used `uint16_t` for `window_size_us`, and keeping the `(uint32_t)` integer promotion typecast for 16-bit big-endian CPUs.

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_scan_aux.c:
```diff
@@ -511,6 +512,22 @@ void ull_scan_aux_setup(memq_link_t *link, struct node_rx_pdu *rx)
+
+	/* Skip reception if invalid aux offset */
+	pdu_us = PDU_AC_US(pdu->len, phy, ftr->phy_flags);
+	if ((aux_offset_us + window_size_us) < (pdu_us + EVENT_MAFS_US)) {
```
The calculation, `(aux_offset_us + window_size_us) < (pdu_us + EVENT_MAFS_US)`, is added twice in this. Would it make sense to add a macro or function for it?

cvinayak committed:
Bluetooth: Controller: Define a macro to validate aux offset value

Define a macro to validate aux offset value as it is checked
both in LLL and ULL execution contexts.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Add missing branch prediction in lll_scan_aux

Add missing branch prediction likely/unlikely hint.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

Thalley commented on subsys/bluetooth/controller/ll_sw/lll_scan_aux.h:
```diff
@@ -4,6 +4,9 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
+#define AUX_OFFSET_IS_VALID(offset_us, win_size_us, pdu_us) \
+		(((offset_us) + (win_size_us)) >= ((pdu_us) + (EVENT_MAFS_US)))
```
```suggestion
#define AUX_OFFSET_IS_VALID(_offset_us, _win_size_us, _pdu_us) \
		(((_offset_us) + (_win_size_us)) >= ((_pdu_us) + (EVENT_MAFS_US)))
```

Suggest to prefix the macro parameters with `_` as the names of the variables are quite common in the controller, and to ensure that we don't accidentally run into any issues with the naming.

mmahadevan108 commented:
@cvinayak , is there an issue this PR should be linked to. No worries if there isn't one.

cvinayak commented:
> @cvinayak , is there an issue this PR should be linked to. No worries if there isn't one.

The commit https://github.com/zephyrproject-rtos/zephyr/commit/3590bd648f8fe39b04d81ee6431037e05f026406 that introduced the regression is not part of a release tag, hence I do not see a reason for a bug report issue. This PR be merged without a bug report issue.
